### PR TITLE
[Sequence Feature Extraction] Add truncation

### LINKED ISSUE #12774: max_length parameter in Wav2Vec2FeatureExtractor doesn't affect
## Environment info

- `transformers` version: 4.8.2
- Platform: Windows-10-10.0.19041-SP0
- Python version: 3.8.5
- PyTorch version (GPU?): 1.9.0+cpu (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: no
- Using distributed or parallel set-up in script?:no

### Who can help
@sgugger

## Information

I'm attempting to receive a features vector of short wav (audio) files using wav2vec by using Hugging Face Transformers.

The problem arises when using:

* [ X] the official example scripts: (give details below)
* [ ] my own modified scripts: (give details below)

The tasks I am working on is:
* [ ] an official GLUE/SQUaD task: (give the name)
* [ X] my own task or dataset: (give details below)

## To reproduce
I ask a question about that in [StackOverflow](https://stackoverflow.com/questions/68430210/how-to-limit-the-size-of-the-features-vector-in-wav2vec)
1.  Two wav files for example are in this [drive ](https://drive.google.com/drive/folders/1j-BNp8D8yN16exgoacgDtDrnMF-jQP91?usp=sharing)
2. extract the feature vector of any file by using a pre-trained wav2vec model with a size limit parameter
3. measure the size of the output  vector
4. surprise...

minimal code example: 
```
import librose
import numpy as np
from transformers import Wav2Vec2FeatureExtractor
input_audio, _ = librosa.load(file,
                                      sr=16000)
        features_with_padding = feature_extractor(input_audio, sampling_rate=16000,
                                return_tensors="np", padding="max_length", max_length=60000).input_values                                
        features_without_padding = feature_extractor(input_audio, sampling_rate=16000,
                                  return_tensors="np", max_length=60000).input_values
        print(features_with_padding.shape, features_without_padding.shape)
```

## Expected behavior
The expected behavior is to get only 60K vector size and not 80-90K vector.
Ideally, I want to get the same size vectors but I don't think that it's possible.

Thanks!

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

This PR adds `truncation` to speech-related feature extractors. It should enable use cases such as: https://github.com/huggingface/transformers/issues/12774

Different to our tokenizers we allow truncation to be just True or False => there is no "truncation strategy". The reason is that for feature extractors the input cannot be a "pair" of input sequences so there is essentially just one use case for truncation that applies to all inputs.

The logic is equivalent to Tokenizers with a small exception in error handling. The differences are shown in example 1. & 2. in the following:

1. truncation=True, no padding, no max_length. IMO this should be an error because it's not clear what should be done here. In Tokenizers we don't throw an error, but simply don't do anything. Throwing an error is better here IMO

```python
from transformers import Wav2Vec2FeatureExtractor, BatchFeature
feat_extractor = Wav2Vec2FeatureExtractor()
dummy_inputs = BatchFeature({"input_values": [[0.1, 0.2, 0.3], [0.1]]})

feat_extractor.pad(dummy_inputs, truncation=True) # -> throws error since `max_length` is not defined 
```

2. truncation=True, "longest" padding, no max_length. IMO this should be an error because it doesn't make sense to "truncate" and to pad to the longest tensor in the batch. If we pad to the longest batch, we can't truncate. In Tokenizers we don't throw an error, but simply don't do anything. Throwing an error is better here IMO

```python
from transformers import Wav2Vec2FeatureExtractor, BatchFeature
feat_extractor = Wav2Vec2FeatureExtractor()
dummy_inputs = BatchFeature({"input_values": [[0.1, 0.2, 0.3], [0.1]]})

feat_extractor.pad(dummy_inputs, truncation=True, padding="longest") # -> throws error since `max_length` is not defined and padding not "max_length"
```

3. truncation=True, "max_length" padding. Here the logic is equivalent to tokenizers

```python
from transformers import Wav2Vec2FeatureExtractor, BatchFeature
feat_extractor = Wav2Vec2FeatureExtractor()
dummy_inputs = BatchFeature({"input_values": [[0.1, 0.2, 0.3], [0.1]]})

feat_extractor.pad(dummy_inputs, truncation=True, max_length=2, padding="max_length") # -> output shape is [2, 2]
```

In addition `pad_to_multiple_of` works correctly and equivalent to the Tokenizers. Tests are added to cover all possible use cases.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/feature_extraction_sequence_utils.py [hunk 1]
```diff
@@ -178,10 +181,18 @@ def pad(
                 processed_features[key] = to_py_obj(value)
 
         # Convert padding_strategy in PaddingStrategy
-        padding_strategy, max_length, _ = self._get_padding_strategies(padding=padding, max_length=max_length)
+        padding_strategy = self._get_padding_strategies(padding=padding, max_length=max_length)
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_sequence_utils.py:184]: We don't modify `max_length` here

File: src/transformers/feature_extraction_sequence_utils.py [hunk 2]
```diff
@@ -107,6 +108,8 @@ def pad(
                   different lengths).
             max_length (:obj:`int`, `optional`):
                 Maximum length of the returned list and optionally padding length (see above).
+            truncation (:obj:`bool`, `optional`):
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_sequence_utils.py:111]: Different to Tokenizers, we just need a True/False boolean here since we don't have input_pairs, ...

File: src/transformers/feature_extraction_sequence_utils.py [hunk 3]
```diff
@@ -107,6 +108,8 @@ def pad(
...
+                Activates truncation to cut input sequences longer than `max_length` to `max_length`.
```
Comment (sgugger) [inline src/transformers/feature_extraction_sequence_utils.py:112]:
```suggestion
                Activates truncation to cut input sequences longer than :obj:`max_length` to :obj:`max_length`.
```

File: src/transformers/feature_extraction_sequence_utils.py [hunk 4]
```diff
@@ -107,6 +108,8 @@ def pad(
                   different lengths).
             max_length (:obj:`int`, `optional`):
                 Maximum length of the returned list and optionally padding length (see above).
```
Comment (sgugger) [inline src/transformers/feature_extraction_sequence_utils.py:110]: This should be adapted to also mention truncation now:
```suggestion
                Maximum length of the returned list and optionally padding or truncation length.
```

File: src/transformers/feature_extraction_sequence_utils.py [hunk 5]
```diff
@@ -278,6 +308,47 @@ def _pad(
...
+        Args:
+            processed_features: Dictionary of input values (`List[float]`) / input vectors (`List[List[float]]`) or batch of inputs values (`List[List[int]]`) / input vectors (`List[List[List[int]]]`)
+            max_length: maximum length of the returned list and optionally padding length (see below)
+            pad_to_multiple_of: (optional) Integer if set will pad the sequence to a multiple of the provided value.
+                This is especially useful to enable the use of Tensor Core on NVIDIA hardware with compute capability
+                >= 7.5 (Volta), or on TPUs which benefit from having sequence lengths be a multiple of 128.
+            truncation: (optional) Activates truncation to cut input sequences longer than `max_length` to `max_length`.
+        """
+        if not truncation:
+            # no truncation
```
Comment (sgugger) [inline src/transformers/feature_extraction_sequence_utils.py:330]: I don't feel this comment adds much, given that the test above says exactly the same thing ;-)

File: src/transformers/feature_extraction_sequence_utils.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/feature_extraction_sequence_utils.py:111]: Why is the default to None then? If it's just True/False, it should default to False and there is no need for a third state.

File: tests/test_sequence_feature_extraction_common.py [hunk 6]
```diff
@@ -205,12 +205,137 @@ def _inputs_are_equal(input_1, input_2):
...
+
+        processed_features = BatchFeature({input_name: speech_inputs})
+
+        # truncate to smallest
+        input_3 = feat_extract.pad(
+            processed_features, padding="max_length", max_length=len(speech_inputs[0]), truncation=True
+        )[input_name]
+        input_4 = feat_extract.pad(processed_features, padding="max_length", max_length=len(speech_inputs[0]))[
+            input_name
+        ]
```
Comment (sgugger) [inline tests/test_sequence_feature_extraction_common.py:237]: Those are pretty hard to read after formatting so the indexing could be done on a separate line.

File: tests/test_sequence_feature_extraction_common.py [hunk 7]
```diff
@@ -205,12 +205,137 @@ def _inputs_are_equal(input_1, input_2):
...
+        input_5 = feat_extract.pad(
+            processed_features,
+            padding="max_length",
+            max_length=len(speech_inputs[0]),
+            return_tensors="np",
+            truncation=True,
+        )[input_name]
+        input_6 = feat_extract.pad(
+            processed_features, padding="max_length", max_length=len(speech_inputs[0]), return_tensors="np"
+        )[input_name]
```
Comment (sgugger) [inline tests/test_sequence_feature_extraction_common.py:252]: Same here.

File: tests/test_sequence_feature_extraction_common.py [hunk 8]
```diff
@@ -205,12 +205,137 @@ def _inputs_are_equal(input_1, input_2):
...
+            max_length=len(speech_inputs[1]),
+            truncation=True,
+            return_tensors="np",
+        )[input_name]
+        input_8 = feat_extract.pad(
+            processed_features, padding="max_length", max_length=len(speech_inputs[1]), truncation=True
+        )[input_name]
+        input_9 = feat_extract.pad(
+            processed_features, padding="max_length", max_length=len(speech_inputs[1]), return_tensors="np"
+        )[input_name]
```
Comment (sgugger) [inline tests/test_sequence_feature_extraction_common.py:274]: Same here.

File: tests/test_sequence_feature_extraction_common.py [hunk 9]
```diff
@@ -205,12 +205,137 @@ def _inputs_are_equal(input_1, input_2):
...
+            max_length=len(speech_inputs[0]),
+            pad_to_multiple_of=pad_to_multiple_of,
+            truncation=True,
+        )[input_name]
+        input_11 = feat_extract.pad(
+            processed_features,
+            padding="max_length",
+            max_length=len(speech_inputs[0]),
+            pad_to_multiple_of=pad_to_multiple_of,
+        )[input_name]
```
Comment (sgugger) [inline tests/test_sequence_feature_extraction_common.py:316]: Same here

**Review by sgugger — APPROVED** [review]
Thanks for adding this new feature! My main comment is on the default for `truncation` (why have None if it's just a True/False flag), the rest are just nits.

File: src/transformers/feature_extraction_sequence_utils.py [hunk 10]
```diff
@@ -278,6 +308,47 @@ def _pad(
...
+            return processed_features
+        elif truncation and max_length is None:
+            raise ValueError(
+                "When setting ``truncation=True``, make sure that ``max_length`` is defined and ``padding='max_length'``"
```
Comment (LysandreJik) [inline src/transformers/feature_extraction_sequence_utils.py:334]: This implies that `padding='max_length'` is required when setting `truncation=True` but that's not the case, is it?

**Review by LysandreJik — APPROVED** [review]
LGTM! Great tests.

File: src/transformers/feature_extraction_sequence_utils.py (hunk 10 above)
Comment (patrickvonplaten) [reply src/transformers/feature_extraction_sequence_utils.py:334]: `max_length` is actually required for truncation (see 3 use cases in the docstring above). We could allow `truncate=True` without any padding which would then truncate everything to the small sample in the batch, but this would be different from how we handle it in the tokenizers -> so I think throwing an error and forcing the user to define the `max_length` is the way to go here. @LysandreJik  What do you think?

Comment (LysandreJik) [reply src/transformers/feature_extraction_sequence_utils.py:333]: Ok yes this makes sense! LGTM
